### ISSUE TITLE
TNeuralFit completes training successfully but reports `TrainingAccuracy`, `ValidationAccuracy`, and `TestAccuracy` as 0.00% even when the network demonstrably learns

### ISSUE THREAD

**Issue by CFTechno** [joaopauloschuler/neural-api#173]
# TNeuralFit Accuracy Metrics Report Zero Percent on Training/Validation/Test Sets

## Description
TNeuralFit completes training successfully but reports `TrainingAccuracy`, `ValidationAccuracy`, and `TestAccuracy` as 0.00% even when the network demonstrably learns (outputs vary appropriately, loss decreases, predictions are non-trivial).

This makes TNeuralFit's accuracy metrics unreliable for monitoring training progress and model evaluation.

### Test Case: Hypotenuse Function (Regression)
Based on the official `examples/Hypotenuse` example pattern:

```pascal
program TestNeuralFitMetrics;
{$mode objfpc}{$H+}

uses
  Classes, SysUtils, Math,
  neuralnetwork, neuralvolume, neuraldatasets, neuralfit;

const
  TRAINING_SAMPLES = 1000;
  VALIDATION_SAMPLES = 100;
  TEST_SAMPLES = 100;

var
  NN: TNNet;
  Trainer: TNeuralFit;
  TrainingPairs, ValidationPairs, TestPairs: TNNetVolumePairList;
  i: integer;
  vOutput: TNNetVolume;

function CreateSimplePairs(Count: integer): TNNetVolumePairList;
var
  i: integer;
  vIn, vOut: TNNetVolume;
  X, Y, Z: single;
begin
  Result := TNNetVolumePairList.Create;
  for i := 0 to Count - 1 do
  begin
    X := Random * 100;
    Y := Random * 100;
    Z := Sqrt(X*X + Y*Y);
    
    vIn := TNNetVolume.Create(2, 1, 1);
    vIn.FData[0] := X / 100;
    vIn.FData[1] := Y / 100;
    
    vOut := TNNetVolume.Create(1, 1, 1);
    vOut.FData[0] := Z / 141.42;
    
    Result.Add(TNNetVolumePair.Create(vIn, vOut));
  end;
end;

begin
  TrainingPairs := CreateSimplePairs(TRAINING_SAMPLES);
  ValidationPairs := CreateSimplePairs(VALIDATION_SAMPLES);
  TestPairs := CreateSimplePairs(TEST_SAMPLES);

  NN := TNNet.Create;
  NN.AddLayer(TNNetInput.Create(2, 1, 1));
  NN.AddLayer(TNNetFullConnectReLU.Create(32));
  NN.AddLayer(TNNetFullConnectReLU.Create(32));
  NN.AddLayer(TNNetFullConnect.Create(1));

  Trainer := TNeuralFit.Create;
  Trainer.InitialLearningRate := 0.00001;
  Trainer.LearningRateDecay := 0;
  Trainer.L2Decay := 0;
  Trainer.Verbose := True;

  WriteLn('Training...');
  Trainer.Fit(NN, TrainingPairs, ValidationPairs, TestPairs, 32, 50);

  WriteLn('Final Training Accuracy: ', FormatFloat('0.00%', Trainer.TrainingAccuracy * 100));
  WriteLn('Final Validation Accuracy: ', FormatFloat('0.00%', Trainer.ValidationAccuracy * 100));
  WriteLn('Final Test Accuracy: ', FormatFloat('0.00%', Trainer.TestAccuracy * 100));
  WriteLn('');
  WriteLn('Sample Output vs Expected:');
  vOutput := TNNetVolume.Create(1, 1, 1);
  for i := 0 to 4 do
  begin
    NN.Compute(TestPairs[i].A);
    NN.GetOutput(vOutput);
    WriteLn(Format('  Output: %.4f, Expected: %.4f',
      [vOutput.FData[0], TestPairs[i].B.FData[0]]));
  end;

  TrainingPairs.Free;
  ValidationPairs.Free;
  TestPairs.Free;
  Trainer.Free;
  NN.Free;
end.
```

## Expected Behavior
After 50 epochs of training on the hypotenuse function:
- `TrainingAccuracy` should be **>50%** (network learned mapping)
- `ValidationAccuracy` should be **>40%** (reasonable generalization)
- `TestAccuracy` should be **>40%** (unseen data performance)

## Actual Behavior
```
Final Training Accuracy: 0.00%
Final Validation Accuracy: 0.00%
Final Test Accuracy: 0.00%

Sample Output vs Expected:
  Output: 0.5585, Expected: 0.8089   (network outputs vary, learning occurred)
  Output: 0.6195, Expected: 0.9582
  Output: 0.3811, Expected: 0.4340
  Output: 0.5848, Expected: 0.8370
```

## Analysis
- ✅ Training **completes successfully** (no crashes)
- ✅ Network **demonstrably learns** (outputs vary, are non-trivial, not random)
- ❌ **Reported accuracies are always 0.00%** across all three sets
- ❌ Makes TNeuralFit's metric reporting **unreliable for monitoring**

## Possible Root Causes
1. `InferHitFn` default implementation incompatible with regression targets
2. Threading (TNeuralFit allocates 32 threads by default) interferes with metric calculation
3. Accuracy calculation expects different target encoding/format
4. Silent failure in metric computation while training proceeds

## Workaround
Use manual training loops with `NN.Compute()` + `NN.Backpropagate()` for reliable accuracy metrics:
```pascal
for epoch := 1 to MAX_EPOCHS do
begin
  for i := 0 to TrainingPairs.Count - 1 do
  begin
    NN.Compute(TrainingPairs[i].A);
    NN.Backpropagate(TrainingPairs[i].B);
  end;
  // Manual accuracy computation via forward passes
end;
```

## Environment
- CAI Neural API: (latest from master branch)
- FreePascal: 3.3.1+
- Platform: Windows 10/11
- Lazarus: 3.x

## Impact
This prevents reliable use of TNeuralFit for model evaluation and progress monitoring, forcing users back to manual training loops even when batching would be beneficial.

**Comment by joaopauloschuler** [comment]
@CFTechno ,
with https://github.com/joaopauloschuler/neural-api/commit/2ff97c581548326d058df89df5ec35d9a11adebb , before fitting, try calling `EnableRegressionComparisonWide()`.

**Comment by joaopauloschuler** [comment]
Does it solve the problem?

**Comment by CFTechno** [comment]
# GitHub Issue #173 - Test Results and Follow-up

## Summary
Tested the recommended fix (`EnableRegressionComparisonWide()`) from commit 2ff97c5. The solution **partially resolves** the issue but leaves test metrics still broken.

## Test Environment
- **Project**: testneuralfit_v2.lpr (copy of original test case)
- **Commit**: 2ff97c5 (Add regression comparison methods and tests for floating-point targets #173)
- **Commit Status**: ✅ **INSTALLED AND VERIFIED**
  - Method confirmed: `procedure EnableRegressionComparisonWide()` at line 1817
  - Implementation: Sets `FInferHitFn := RegressionCompareWide`
  - Also includes `EnableRegressionComparison()` for regular regression
- **Fix Applied**: `Trainer.EnableRegressionComparisonWide()` called before `Fit()`
- **Task**: Hypotenuse regression (x² + y² → √(x²+y²))
- **Configuration**: 
  - Network: Input(2) → ReLU(32) → ReLU(32) → FC(1)
  - Training: 50 epochs, batch size 32, LR 0.00001
  - Datasets: 1000 training, 100 validation, 100 test samples

## Results

### Before Fix (Original testneuralfit)
```
Final Training Accuracy: 0.00%
Final Validation Accuracy: 0.00%
Final Test Accuracy: 0.00%
```

### After Fix (testneuralfit_v2 with EnableRegressionComparisonWide)
```
Final Training Accuracy: 99.91%
Final Validation Accuracy: 100.00%
Final Test Accuracy: 0.00%  ❌ STILL BROKEN
```

### Detailed Output
```
Test TNeuralFit on Hypotenuse (Regression Task)
==============================================

Creating datasets...
  Training: 1000 samples
  Validation: 100 samples
  Test: 100 samples

Creating neural network...
  Network created: 65 neurons

Configuring TNeuralFit...
  Learning Rate: 0.00001
  Batch Size: 32
  Regression Comparison: Enabled

Starting training (50 epochs)...
[Training output truncated]

Training complete!
Elapsed time: 0.05 seconds

Final Accuracies:
  Training: 99.91%
  Validation: 100.00%
  Test: 0.00%  ← Still reporting 0.00%

Sample predictions (first 5 test samples):
  Input: 0.54, 0.34 -> Output: -0.0358 (Expected: 0.4529)
  Input: 0.48, 0.68 -> Output: 0.0063 (Expected: 0.5898)
  Input: 0.38, 0.03 -> Output: -0.0452 (Expected: 0.2726)
  Input: 0.71, 0.50 -> Output: -0.0415 (Expected: 0.6144)
  Input: 0.09, 0.59 -> Output: 0.0055 (Expected: 0.4248)
```

## Analysis

### ✅ What's Fixed
- **Training accuracy now reports correctly**: 99.91% (was 0.00%)
- **Validation accuracy now reports correctly**: 100.00% (was 0.00%)
- `EnableRegressionComparisonWide()` successfully enables metric computation for training and validation sets

### ❌ What's Still Broken
- **Test accuracy still reports 0.00%** despite having 100 test samples
- The test set metrics are not being computed even with the fix enabled
- This suggests the fix is incomplete or test metrics require additional configuration

### 🔍 Secondary Issue Observed
- Network shows extreme overfitting: 99.91% training / 100% validation but terrible test predictions
- This is unusual for such a simple task and may indicate:
  - The metrics calculation method doesn't align with actual network performance
  - Or the network genuinely overfitted to the train/val sets

## Questions for @joaopauloschuler

1. **Is test metric computation expected to work with `EnableRegressionComparisonWide()`?**
   - Or does the test set require separate handling (e.g., `EnableRegressionComparisonWideTest()`)?

2. **Is there additional configuration needed** for the test set metrics to be computed?

3. **Are training/validation metrics now considered "solved"** with this commit, with test metrics as a known limitation?

## Code Used in Test

```pascal
// Configure TNeuralFit
Trainer := TNeuralFit.Create;
Trainer.InitialLearningRate := 0.00001;
Trainer.LearningRateDecay := 0;
Trainer.L2Decay := 0;
Trainer.Verbose := True;
// Enable regression comparison for floating-point targets (fix for GitHub issue #173)
Trainer.EnableRegressionComparisonWide;

// Train
Trainer.Fit(NN, TrainingPairs, ValidationPairs, TestPairs, 32, 50);

// Check metrics
WriteLn('Training: ', FormatFloat('0.00%', Trainer.TrainingAccuracy * 100));
WriteLn('Validation: ', FormatFloat('0.00%', Trainer.ValidationAccuracy * 100));
WriteLn('Test: ', FormatFloat('0.00%', Trainer.TestAccuracy * 100));
```

## Recommendation

The fix is **a positive step forward** but appears to be incomplete:
- ✅ Solves the primary issue for training and validation metrics
- ⚠️ Test metrics remain unresolved

Suggest checking if:
1. Test metrics computation is implemented in the commit
2. Whether a follow-up fix is needed for test set metrics
3. Or if test metrics are out of scope for this issue

**Comment by achinastone** [comment]
@DrDub 
// Enable regression comparison for floating-point targets (fix for GitHub issue #173)
Trainer.EnableRegressionComparisonWide;
Trainer.TargetAccuracy := 2
